src/prog_models/lstm_model.py

```diff
@@ -0,0 +1,387 @@
+        if not isinstance(data, Iterable):
+            raise ValueError(f"data must be in format [(input, output), ...], got {type(data)}")
+        if len(data) == 0:
+            raise ValueError("No data provided. Data must be in format [(input, output), ...] and have at least one element")
+        if not isinstance(data[0], tuple):
+            raise ValueError(f"Each element of data must be a tuple, got {type(data[0])}")
+        if len(data[0]) != 2:
+            raise ValueError("Each element of data must be in format (input, output), where input and output are either np.array or SimulationResults and have at least one element")
+        if not isinstance(params['normalize'], bool):
+            raise TypeError(f"normalize must be a boolean, not {type(params['normalize'])}")
+
+        # Prepare datasets
+        (u_all, z_all) = LSTMStateTransitionModel.pre_process_data(data, **params)
+
+        # Normalize
+        if params['normalize']:
+            n_inputs = len(data[0][0][0])
+            u_mean = np.mean(u_all[:,0,:n_inputs], axis=0)
+            u_std = np.std(u_all[:,0,:n_inputs], axis=0)
+            # If there's no variation- dont normalize 
+            u_std[u_std == 0] = 1
+            z_mean = np.mean(z_all, axis=0)
+            z_std = np.std(z_all, axis=0)
+            # If there's no variation- dont normalize 
+            z_std[z_std == 0] = 1
+
+            # Add output (since z_t-1 is last input)
+            u_mean = np.hstack((u_mean, z_mean))
+            u_std = np.hstack((u_std, z_std))
+
+            u_all = (u_all - u_mean)/u_std
+            z_all = (z_all - z_mean)/z_std
+
+            # u_mean and u_std act on the column vector form (from inputcontainer)
+            # so we need to transpose them to a column vector
+            params['normalization'] = (u_mean[np.newaxis].T, u_std[np.newaxis].T, z_mean, z_std)
+        
+        # Build model
+        callbacks = [
+            keras.callbacks.ModelCheckpoint("jena_sense.keras", save_best_only=True)
+        ]
+
+        inputs = keras.Input(shape=u_all.shape[1:])
+        x = inputs
+        for i in range(params['layers']):
+            if i == params['layers'] - 1:
+                # Last layer
+                x = layers.LSTM(params['units'][i], activation=params['activation'][i])(x)
+            else:
+                # Intermediate layer
+                x = layers.LSTM(params['units'][i], activation=params['activation'][i], return_sequences=True)(x)
+        
+        if params['dropout'] > 0:
+            # Dropout prevents overfitting
+            x = layers.Dropout(params['dropout'])(x)
+
+        x = layers.Dense(z_all.shape[1] if z_all.ndim == 2 else 1)(x)
+        model = keras.Model(inputs, x)
+        model.compile(optimizer="rmsprop", loss="mse", metrics=["mae"])
+        
+        # Train model
+        model.fit(u_all, z_all, epochs=params['epochs'], callbacks = callbacks, validation_split = params['validation_split'])
```

Comment:
I assume based of the last meeting that this fitness fucntion will be changed. It had something to do with seperate the parts of LSTM so as to not impact and cause the jumpiness we all saw in the output?